rfd/0061-tsh-aliases.md

```diff
@@ -0,0 +1,166 @@
+---
+authors: Roman Tkachenko (roman@goteleport.com)
+state: draft
+---
+
+# RFD 61 - tsh command aliases
+
+## What
+
+Proposes a way for users to define custom tsh commands.
+
+## Why
+
+Defining custom commands will allow users to extend `tsh` CLI to fit their
+use-case without having to modify Teleport's source code.
+
+Examples of what users might want to do include:
+
+- Define shorthands for the existing `tsh` commands.
+
+- Tweak the default behavior of existing `tsh` commands, for example to have
+  `tsh login` always default to a particular auth connector or username.
+
+- Define new subcommands combining multiple actions in the same command, for
+  example logging into a cluster, a database and connecting to the database.
+
+## Prior art
+
+### bash aliases
+
+The simplest analogue that comes to mind is the bash alias:
+
+```bash
+alias ll="ls -l"
+```
+
+Bash aliases are quite simplistic as they only allow to define new commands
+that expand to the strings they alias and don't provide a way to refer to the
+command's arguments.
+
+### git aliases
+
+Git aliases are defined in the repo-specific `.gitconfig` (or `.git/config`)
+or global `/etc/gitconfig` file, and are much closer conceptually to what we
+want to accomplish as they allow to define custom `git` subcommands:
+
+```ini
+[alias]
+co = checkout
+ls = log --oneline
+```
+
+Git aliases also support calling non-`git` commands (using `!` notation) and
+refer to the command's arguments (via `$@`, `$1`, etc.) which allows to build
+powerful new commands.
+
+For example, the following alias will make `git top` to show 15 most recently
+updated local branches:
+
+```ini
+[alias]
+top = "!bash -c 'git branch --sort=-committerdate | head -15'"
+```
+
+## Configuration and examples
+
+Users will define aliases in their `tsh` configuration file which is kept
+in `$TELEPORT_HOME/config/config.yaml` using the following syntax:
```

Comment:
It may be useful to support a global `tsh` configuration file `/etc/tshconfig` (as mentioned in prior art).

Reply:
@r0mant goot point, see above.

Reply:
Yep, I also thought about this, just wasn't sure how often it is that `tsh` is used by multiple different users on the same machine (which is the primary use-case for this global config IIUC).

I've added a section to the RFD about it, we can make the global config merge with the local config, with local config taking precedence for any conflicting settings.